sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +315,191 @@ case class GpuCaseWhen(
+                }
+              }
+          }
+        }
+      }
+    } finally {
+      currentValue.foreach(_.safeClose())
+      cumulativePred.foreach(_.safeClose())
+    }
+  }
+
+  /**
+   * Filter the batch to just the rows where the WHEN condition is true and
+   * then evaluate the THEN expression.
+   */
+  private def filterEvaluateWhenThen(
+      colTypes: Array[DataType],
+      tbl: Table,
+      whenBool: ColumnVector,
+      thenExpr: Expression): ColumnVector = {
+    val filteredBatch = filterBatch(tbl, whenBool, colTypes)
+    val thenValues = withResource(filteredBatch) { trueBatch =>
+      GpuExpressionsUtils.columnarEvalToColumn(thenExpr, trueBatch)
+    }
+    withResource(thenValues) { _ =>
+      gather(whenBool, thenValues)
+    }
+  }
+
+  /**
+   * Calculate the cumulative predicate so far using the logical expression
+   * `prevPredicate OR thisPredicate`.
+   */
+  private def calcCumulativePredicate(
+      cumulativePred: Option[GpuColumnVector],
+      whenBool: GpuColumnVector,
+      firstTrueWhen: GpuColumnVector): GpuColumnVector = {
+    cumulativePred match {
+      case Some(prev) =>
+        withResource(prev) { _ =>
+          val result = withResource(new Table(prev.getBase, whenBool.getBase)) { t =>
+            val or = new ast.BinaryOperation(ast.BinaryOperator.NULL_LOGICAL_OR,
+              new ast.ColumnReference(0),
+              new ast.ColumnReference(1)
+            )
+            withResource(or.compile()) {
+              _.computeColumn(t)
+            }
+          }
+          GpuColumnVector.from(result, DataTypes.BooleanType)
+        }
+      case _ =>
+        firstTrueWhen.incRefCount()
+    }
+  }
+
+  /**
+   * Calculate the current values by merging the THEN values for this branch (where the WHEN
+   * predicate was true) with the previous values.
+   */
+  private def calcCurrentValue(
+      prevValue: Option[GpuColumnVector],
+      whenBool: GpuColumnVector,
+      thenValues: ColumnVector): GpuColumnVector = {
+    prevValue match {
+      case Some(v) =>
+        withResource(v) { _ =>
+          GpuColumnVector.from(whenBool.getBase.ifElse(thenValues, v.getBase), dataType)
+        }
+      case _ =>
+        GpuColumnVector.from(thenValues.incRefCount(), dataType)
+    }
+  }
+
+  /**
+   * Determine for each row whether this is the first WHEN predicate so far to evaluate to true
+   */
+  private def isFirstTrueWhen(
+      cumulativePred: Option[GpuColumnVector],
+      whenBool: GpuColumnVector): GpuColumnVector = {
+    cumulativePred match {
+      case Some(prev) =>
+        val notPreviouslyTrue = withResource(prev.getBase.not()) { previouslyFalse =>
+          withResource(prev.getBase.isNull) { previouslyNull =>
+            previouslyFalse.or(previouslyNull)
+          }
+        }
+        val result = withResource(new Table(whenBool.getBase, notPreviouslyTrue)) { t =>
+          val expr = new ast.BinaryOperation(ast.BinaryOperator.NULL_LOGICAL_AND,
```

Comment:
This won't work because `whenBool` can contain nulls, and null AND something is always null, even with the AST LOGICAL_AND.  I think we can avoid AST and leverage what we've already done previously by doing something like this.  boolInverted has the nice property of treating nulls like they were originally false.
```scala
  withResource(boolInverted(prev.getBase)) { notPrev =>
    val whenReplaced = withResource(Scalar.fromBool(false)) { falseScalar =>
      whenBool.getBase.replaceNulls(falseScalar)
    }
    GpuColumnVector.from(whenReplaced.and(notPrev), DataTypes.BooleanType)
  }
```

Reply:
Thanks. I added tests to hit the cases where predicates can evaluate to null and confirmed that they failed with the original code and pass with the latest code (including the above suggestion).